### PR TITLE
[MAD-16773] Use poll() instead select() to avoid Android SIGABORT crash on socket state polling

### PR DESCRIPTION
**Pull request by vmednonogov**
Ticket: 16773

## What does this PR do?
Issue: the game crashes with SIGABORT on android when fd_set() and select() want to use the sockets with descriptor values >= FD_SETSIZE

Changes: use poll() instead select() for Android to allow the sockets with descriptor values >= FD_SETSIZE
Based on this article: https://www.ulduzsoft.com/2014/01/select-poll-epoll-practical-difference-for-system-architects/

Note1: FTUE and WorldMap start and work after this fix.
 But battle  crashes by the same reason but on another place. Created PR to fix it in game code https://github.com/carbonated-dev/o3de-gruber/pull/1169.

Note2: this problem is absent on PC/Mac/iOS, looks like it was fixed in native libs. But solution with poll() should work on PC/Mac/iOS too (on PC we need to use WSAPoll() instead poll())

Note3: why we did not see this issue before - I don't know. For example we could use less sockets, the some first sockets are created with small descriptor values (valid for Android).

## How was this PR tested?

PC/Android locally

Also verified build 28664 apk:
 MW_build_vmedn_MAD_16773_use_poll_instead_select_to_avoid_android_crash_0.8.2.28664.qa.apk
 (Visited Tut, RV, Map, played battle)

### CODE REVIEW

**Review by Twolewis — APPROVED** [review]

**Comment by Twolewis** [comment]
@vmednonogov  - did you discuss w/ o3de folks?  This seems like a good one to get into O3DE proper

**Comment by vmednonogov** [comment]
> @vmednonogov - did you discuss w/ o3de folks? This seems like a good one to get into O3DE proper

I will share this information with o3de team in Slack, when finished to verify this PR via Jenkins

**Comment by vmednonogov** [comment]
Verified build 28664 apk:
 MW_build_vmedn_MAD_16773_use_poll_instead_select_to_avoid_android_crash_0.8.2.28664.qa.apk
 (Visited Tut, RV, Map, played battle)

File: Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/Socket/AzSocket_UnixLike.cpp [hunk 1]
```diff
@@ -294,8 +294,24 @@ namespace AZ::AzSock
...
     {
+#if defined(CARBONATED) && defined(AZ_PLATFORM_ANDROID)
+        // Use poll() instead select() for Android to allow the sockets with descriptor values >= FD_SETSIZE
+        AZPOLLFD fdPollData;
+        fdPollData.fd = sock;
+        fdPollData.events = POLLIN;
+        fdPollData.revents = 0;
+
+        return Poll(&fdPollData, 1, timeout ? timeout->tv_sec * 1000 + timeout->tv_usec / 1000 : -1) > 0 ? 1 : 0;
+#else
```
Comment (Twolewis) [inline Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/Socket/AzSocket_UnixLike.cpp:314]: Forgot to mention - O3DE really tries to minimize platform-specific guards like AZ_PLATFORM_ANDROID.  Given this is socket code, is it a change we can enable for all unix-like platforms?  Will this work for iOS & Linux?

If so, we should ditch the AZ_PLATFORM_ANDROID guards.  Also please make a PR back to O3DE.

Comment (vmednonogov) [reply Code/Framework/AzCore/Platform/Common/UnixLike/AzCore/Socket/AzSocket_UnixLike.cpp:314]: Added a new trait for socket polling in new PRs:
https://github.com/carbonated-dev/o3de/pull/348
https://github.com/carbonated-dev/o3de-gruber/pull/1176

**Pull request merged**